mapper/collectd_mapper/src/functional_batcher.rs

```diff
@@ -0,0 +1,345 @@
+//! Batching algorithm that is unaware of IO. Imperative shell, functional core.
+
+use crate::collectd::OwnedCollectdMessage;
+use clock::Timestamp;
+
+#[derive(Debug, PartialEq)]
+pub struct MessageBatch(pub Vec<OwnedCollectdMessage>);
+
+/// The Batcher's internal state.
+pub struct Batcher {
+    /// Maximum number of messages per batch.
+    max_batch_size: usize,
+
+    /// The maximum age of a batch.
+    ///
+    /// Age of a batch is the elapsed time since `current_batch.opened_at`.
+    max_batch_age: chrono::Duration,
+
+    /// We start a new batch upon receiving a message whose timestamp is farther away to the
+    /// timestamp of first messsge in the batch than `collectd_timestamp_delta` seconds.
+    ///
+    /// Delta is inclusive.
+    collectd_timestamp_delta: f64,
+
+    current_batch: CurrentBatch,
+}
+
+// Invariants:
+//
+//   opened_at.is_none() => messages.is_empty().
+//   opened_at.is_some() => messages.len() > 0.
+//
+struct CurrentBatch {
+    opened_at: Option<Timestamp>,
+    messages: Vec<OwnedCollectdMessage>,
+}
+
+impl CurrentBatch {
+    fn empty() -> Self {
+        Self {
+            opened_at: None,
+            messages: Vec::new(),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.messages.is_empty()
+    }
+}
+
+/// Inputs to the `Batcher`.
+pub enum Input {
+    /// A message was received.
+    Message {
+        /// Time the message has been received.
+        received_at: Timestamp,
+
+        /// The message itself.
+        message: OwnedCollectdMessage,
+    },
```

Comment:
I don't completely follow the need for two timestamps here.

* I guess one is the reception time (by the batcher) and the other is the event time (assigned by collectd).
* I feel that this message is a mix of two input events: a new message and a tick.

Reply:
Yes, I see your point. Think about the `Tick` more as a "firing of a timer" than as the "tick" of a clock. Then it should make more sense. There won't be a Tick for every message informing the algorithm that time has advanced. 

Probably needs some renaming of `Tick` to `TimerFired` and `NextTickAt` to `LoadTimer`.